### PR TITLE
Part of #1475 - Enable training of multi-heads model

### PR DESCRIPTION
**Pull request by sjhpark**
This revision of the Ultrallytics library allows multi-heads training from its dedicated classes (Model, Loss, Detect, Pose).
The assigned name for this multi-heads training task is "pose-multiclsheads".

==== Test ====
`./start.sh <dataset dir> <save_path> <task> shell -u <local ultralytics dir to mount> -c <name of container to create>`

- I trained a model with pose-multiclsheads task for 100 epochs on a merged dataset.
- I modified the saved onnx file of this trained model to select only one classification head.
- I evaluated it on an out-of-distribution (not seen by the model) data.

More details: https://www.notion.so/Training-19383cede6a5806fba09c46d464bcf0f?pvs=4

### CODE REVIEW

File: ultralytics/utils/loss.py [hunk 1]
```diff
@@ -772,6 +772,100 @@ def select_rand_idx(n_samples, targets, class_label):
...
+        # gt_labels: (B, S, 1) where S is # of max targets (gt bboxes) in any image in the batch
+        # extract field classes from gt_labels       
+        if gt_labels.shape[1] == 0:  # if there are no targets across all field images in the batch
+            # multiply 0 to pred scores to not affect training while allowing gradient computation during GradScaler's backprop
+            zero_loss = abs(torch.zeros(self.n_losses, device=self.device)  * pred_scores.sum())
+            return zero_loss.sum() * batch_size, zero_loss.detach()
+        
+        # assign a crop field class to each image in the batch
+        # (we reason on the very first target's class to determine the field for the image)
+        field_classes = gt_labels[:, 0].int().squeeze() // 2  # classes (crop & weed) per crop field image (e.g., [0,1,2,3] -> [0,0,1,1]); (B,)
```
Comment (romainthibaux) [inline ultralytics/utils/loss.py:818]: Here we shouldn't need to hardcode again that the heads are all of size 2. We should be able to look at the channel size of each head in order and add up the channel sizes to understand which label is associated with which head.

File: ultralytics/utils/loss.py [hunk 2]
```diff
@@ -772,6 +772,100 @@ def select_rand_idx(n_samples, targets, class_label):
...
+
+        # 2. convert gt_labels = gt_labels % 2 (e.g., [0,1,2,3] -> [0,1,0,1])
+        gt_labels = gt_labels % 2  # gt_labels[:, :] % 2
```
Comment (romainthibaux) [inline ultralytics/utils/loss.py:821]: Same. Also delete the comment.

File: ultralytics/utils/loss.py [hunk 3]
```diff
@@ -772,6 +772,100 @@ def select_rand_idx(n_samples, targets, class_label):
...
+
+        # Pboxes
+        pred_bboxes = self.bbox_decode(anchor_points, pred_distri)  # xyxy, (b, h*w, 4)
+        pred_kpts = self.kpts_decode(anchor_points, pred_kpts.view(batch_size, -1, *self.kpt_shape))  # (b, h*w, 17, 3)
+
+        # slice pred_scores
+        # For each batch (B) & all anchors (N), drop all irrelevant classes (C) from classification predictions
+        for b, c in enumerate(field_classes):
+            c_offset = c * 2  # 0 -> 0, 1 -> 2, 2 -> 4, ... where label 2 will actually mean class 4 (crop) or 5 (weed) in the crop field C
+            pred_scores[b, :, 0:2] = pred_scores[b, :, c_offset:c_offset+2]  # bring classes (crop & weed for the field) to the front
```
Comment (romainthibaux) [inline ultralytics/utils/loss.py:831]: We should not mention crop and weed in this code. This code doesn't care about the application.

File: ultralytics/utils/loss.py [hunk 4]
```diff
@@ -772,6 +772,100 @@ def select_rand_idx(n_samples, targets, class_label):
...
+            anchor_points * stride_tensor, gt_labels, gt_bboxes, mask_gt)
+
+        # slice target scores as prediction scores are sliced
+        target_scores = target_scores[:, :, 0:2]
+
+        target_scores_sum = max(target_scores.sum(), 1)
+
+        # Cls loss
+        # loss[1] = self.varifocal_loss(pred_scores, target_scores, target_labels) / target_scores_sum  # VFL way
+        loss[3] = self.bce(pred_scores, target_scores.to(dtype)).sum() / target_scores_sum  # BCE
```
Comment (romainthibaux) [inline ultralytics/utils/loss.py:846]: Why the commented out code?

Comment (sjhpark) [reply ultralytics/utils/loss.py:846]: That commented out code was carried from the original Ultralytics library's DetectionLoss and PoseLoss code in `/utils/loss.py`
https://github.com/ultralytics/ultralytics/blob/efc3c392a9441e32171536520e39fa5ce7413916/ultralytics/utils/loss.py#L246
I will remove it from my class.

File: ultralytics/utils/loss.py [hunk 5]
```diff
@@ -772,6 +772,100 @@ def select_rand_idx(n_samples, targets, class_label):
...
+        # shift gt_labels to 0 (crop) and 1 (weed) for binary classification
+        # pred_scores: (B, N, C) where B is batch size, N is num anchors, C is num classes
+        # gt_labels: (B, S, 1) where S is # of max targets (gt bboxes) in any image in the batch
+        # extract field classes from gt_labels       
+        if gt_labels.shape[1] == 0:  # if there are no targets across all field images in the batch
+            # multiply 0 to pred scores to not affect training while allowing gradient computation during GradScaler's backprop
+            zero_loss = abs(torch.zeros(self.n_losses, device=self.device)  * pred_scores.sum())
+            return zero_loss.sum() * batch_size, zero_loss.detach()
+        
+        # assign a crop field class to each image in the batch
```
[the remark this answers is not in the document]
Comment (sjhpark) [reply ultralytics/utils/loss.py:816]: I replaced "field_class" with "head_class" and avoided using the terms "crop", "weed", and "field".

File: ultralytics/utils/loss.py (hunk 2 above)
Comment (sjhpark) [reply ultralytics/utils/loss.py:821]: Removed as suggested.

File: ultralytics/utils/loss.py (hunk 3 above)
Comment (sjhpark) [reply ultralytics/utils/loss.py:831]: Removed "crop" and "weed" from the comments.

File: ultralytics/utils/loss.py (hunk 1 above)
Comment (sjhpark) [reply ultralytics/utils/loss.py:818]: Currently there is no information about the out_channel size of each convolution head is passed to this loss function.
I need to think of a way to do this.

I think your idea is still using "2" classes per head but just inferring from a channel size of a head, right?
Otherwise, I am currently slicing pred_scores (batch, num_anchors, num_cls) with hard coded "2" to select 2 classes, and pred_scores's batches are randomly selected images, so I need to find a way to match the images with corresponding classification heads.
